### PR TITLE
metrics-loader: create metrics-loader tool

### LINKED ISSUE #130403: roachtest: write OpenMetrics to Prometheus Remote Write endpoint
We require a way to load `OpenMetrics` [1] files into compatible long-term metric storage systems. Prometheus defines an API spec [2] which is implemented by multiple vendors. The spec describes an endpoint that can consume metrics via HTTP, which can serve as a way to bulk load metrics if the underlying system is compatible with backfilling, such as [3] (`promtable` uses `Bigtable` as a metric storage medium and supports both the Prometheus remote read and write protocol).

With the implementation of #106662, which allows us to produce OpenMetrics files from `roachtest` metrics, this issue aims to address the concern of loading these files into a suitable metrics storage. The Prometheus write API is a good standard to target since many systems already support this.

This issue should provide a solution for reading OpenMetrics file and writing the metrics to a Prometheus Write endpoint according to the spec [2]. The solution should also take into account that we produce multiple files each night, with the nightly runs. This means there could be infra flakes between the producer and consumer. It may be good idea to decouple these systems.

[1] https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md
[2] https://prometheus.io/docs/specs/remote_write_spec/#prometheus-remote-write-specification
[3] https://github.com/uschen/promtable

Jira issue: CRDB-42037

### PR DESCRIPTION
**Pull request by herkolategan**
This PR contains several commits to add a new tool to the arsenal, the `metrics-loader`.
The main purposes of the tool is to monitor a Google Storage Bucket (path) for OpenMetrics format files [1] and stream those metrics to a Prometheus remote-write endpoint [2].

The tool is built up out of several components. Each component's goal is described in more detail in the commit message.
A brief overview of all the components:
- Reader: Decodes an OpenMetrics file into Prometheus metric structs.
- Sender: Sends `decoded` metrics to a Prometheus remote-write endpoint
- Processor: Process a given input (File) by utilising a `Reader` &`Sender`.
- Storage: A simple layer to interact with a storage system (GCS in this case)
- Watcher: Watches the Google Bucket for changes, and invokes the Processor to process any new files.

[1] https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md?rgh-link-date=2024-09-10T10%3A57%3A25Z
[2] https://prometheus.io/docs/specs/remote_write_spec/#prometheus-remote-write-specification

Resolves: #130403
Informs: #106662

Epic: None
Release note: None

### CODE REVIEW

**Review by DarrylWong — COMMENTED** [review]
Should we add this to dev build? 

I tried adding something like:

 `"metrics-loader":      "//pkg/cmd/metrics-loader:metrics-loader",`
 
 [to the mapping](https://github.com/cockroachdb/cockroach/blob/41c3aa4084245e319ad92128f594706ea5fdc77f/pkg/cmd/dev/build.go#L78) but not sure what else needs to be done to get it to work.

**Comment by DarrylWong** [comment]
Should we also add this new path to `.github/CODEOWNERS` under test-eng?

File: pkg/cmd/metrics-loader/watcher.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+	reader, err := w.googleStorage.Read(uri)
+	if err != nil {
+		return err
+	}
+	defer reader.Close()
+	err = w.processor.Process(reader)
+	if err != nil {
+		return err
+	}
+	return w.googleStorage.Move(uri, w.toProcessedPath(uri))
```
Comment (DarrylWong) [inline pkg/cmd/metrics-loader/watcher.go:148]: If `Process` already sends the metrics to prom, whats the reason for keeping the file around as opposed to deleting it?

**Comment by DarrylWong** [comment]
On second thought, I wonder if this should live in its own repo. I don't see why this needs to live in the monorepo besides convenience but perhaps that's good enough reason 🤷 

Overall this looks good on first pass, but I haven't gotten around to actually trying it out myself.

File: pkg/cmd/metrics-loader/reader.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+type openMetricsReader struct{}
+
+func NewOpenMetricsReader() Reader {
+	return &openMetricsReader{}
+}
+
+func (r *openMetricsReader) Read(reader io.Reader, outputFn ReaderOutputFn) error {
+	batch := make([]byte, 0, readBufferSize+readBatchSize)
```
[the remark this answers is not in the document]
Comment (herkolategan) [reply pkg/cmd/metrics-loader/reader.go:41]: Good catch! That was based on something else originally, I'll amend!

**Comment by herkolategan** [comment]
> On second thought, I wonder if this should live in its own repo.
Good question, since it doesn't really have any dependencies in the main repo I'm not opposed to moving it under a new repo within cockroachdb. I just want to move it out of my own GH so others can contribute.

**Comment by herkolategan** [comment]
Moving this to a different repo.

**Pull request closed without merging**